### PR TITLE
Revolver Heavy Ammo Effect Change

### PR DESCRIPTION
**Pull request by Steelpoint**
# About the pull request

This PR changes heavy ammo for the Revolver to knockback a mob and slow them down instead of stunning it. 

# Explain why it's good for the game

Combat balance is a precarious and often difficult conversation to hold, ergo I'll lay my biases out on the table at first.  I'm a marine main at heart, but I have played a lot of xeno recently to gain a better understanding of their side of the story, enough that I feel confident to make these assertions.

My belief is that the heavy ammo of the revolver is a negative concept for the game, and it needs to be removed, due to its stun factor. 

The issue here is readability and prediction. When you see a RPG, you know that it can fire a devastating warhead that can stun and kill T3s. When you see a Warrior, you know it can leap to 4 tiles to stun and drag a Human, when you hear a CAS strike you know exactly what is about to drop. When you see a Queen you know she can stun screech and neuro stun you. 

But the issue with the Revolver is it has no obvious tell. It is a small item, that can be fit inside backpacks, holsters, pouches, belts, armour slots. It has no obvious advance warning when you are going to fire it. There is no special uniform requirement making a revolver user standout amongst the crowd. There is no tell.

The problem with the stun revolver is simply that is is a hard counter to all T1s and most T2s. Its ability to stun allows it to perform an attack that is uncounterable to a xeno as a xeno has no way to predict who may be carrying one. A xeno can tell who a Specialist is, a xeno can tell who has a shotgun or flamer or sniper or RPG, you can tell when a mortar is being prepared, or a CAS strike or even an OB. You can see the smartgunner. Even the Scout, a literal cloaked Marine, has to uncloak to fire. You can not tell who has a revolver until they pull it out and stun you. Once you are stunned you die. 

A xeno equilivant would be if any Xeno could be carrying a special tool that lets them grab a marine from 7 tiles away and pull them in plus stun them for 2 seconds. But any xenomorph could be using it, including a Lesser Drone. 

Perhaps the heavy revolver could be reworked to do something else, but ultimately the only reason anyone takes this ammo is for the stun. Anything else is beating around the bush. 

Those are my reasoning's, I'll leave the rest to the powers' that be. 

# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog
:cl:
balance: Revolver Heavy ammo no longer stuns targets it strikes, it will instead knock them back and slow them down for a short time. 
/:cl:

### CODE REVIEW

**Comment by Steelpoint** [comment]
I could alter the ammo to pushback a mob instead of stunning, so it could still be useful as an escape tool to waylay or stop an aggressor but it would not stunlock a target.

**Comment by Drulikar** [comment]
> I could alter the ammo to pushback a mob instead of stunning, so it could still be useful as an escape tool to waylay or stop an aggressor but it would not stunlock a target.

Go for it - a revolver being a defensive weapon instead of an offensive weapon sounds more in line anyways.

**Comment by AGasCanister** [comment]
stopkillingdevsguys

**Comment by Steelpoint** [comment]
I've returned the heavy revolver ammo availability to what it currently uses, but have changed the bullet to knockback a mob that is struck within 4 tiles of the firer's position.

**Review by Drulikar — APPROVED** [review]
Code looks fine and works, just up to Zonespace if they think this change fits balance.

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by Steelpoint** [comment]
Merge conflicts resolved.

**Comment by Drulikar** [comment]
For future reference, resolve conflicts by merging upsteam master into your branch. No force pushing is necessary.

**Comment by 13lacklight** [comment]
just limit the stuns range, a push back is useless as a defensive weapon cause the xeno will just come running back. half the point of heavy ammo is that it makes attacking backline support personnel a little dicier rather than a near auto win.
shooting a runner doesnt spook it cause it knows it can just run you down anyway. stun it, and it shits itself. hit it with knock back and half a second later its exactly where it was and youre out of ammo.
If youre gonna outright remove the stun, then maybe we should remove shotguns as well, cause people can pull them out of any number of slots to PB xenos with little to no warning. or remove slug stun, cause again you can pull it out of no where and stun lock a xeno. 

this just reads like you died to heavy ammo once and salted. not going to lie.

**Review by Zonespace27 — CHANGES_REQUESTED** [review]
Add a short slow on hit and this is good to go

**Comment by Steelpoint** [comment]
I've updated the bullets to now deal slowdown to targets struck. A combination of being pushed back and slowed for 2 seconds should help someone escape an attacker.

File: code/datums/ammo/bullet/revolver.dm [hunk 1]
```diff
@@ -27,8 +27,9 @@
 	penetration = ARMOR_PENETRATION_TIER_4
 	accuracy = HIT_ACCURACY_TIER_3
 
-/datum/ammo/bullet/revolver/heavy/on_hit_mob(mob/M, obj/projectile/P)
-	knockback(M, P, 4)
+/datum/ammo/bullet/revolver/heavy/on_hit_mob(mob/entity, obj/projectile/bullet)
+	slowdown(entity, bullet, 4)
```
Comment (Zonespace27) [inline code/datums/ammo/bullet/revolver.dm:31]: Slowdown doesn't take 3 args

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: code/datums/ammo/bullet/revolver.dm (hunk 1 above)
Comment (Steelpoint) [reply code/datums/ammo/bullet/revolver.dm:31]: I don't understand this statement.

Comment (Drulikar) [reply code/datums/ammo/bullet/revolver.dm:31]: `/datum/ammo/proc/slowdown(mob/living/living_mob, obj/projectile/fired_projectile)`

Argument 1 is living_mob, expected to be a /mob/living; you pass entity to be this argument
Argument 2 is fired_projectile, expected to be a /obj/projectile/; you pass bullet to be this argument
Argument 3 does not exist, but you pass 4 into a non-existent argument.

**Comment by Steelpoint** [comment]
I believe I fixed the outstanding issue

File: code/datums/ammo/bullet/revolver.dm [hunk 2]
```diff
@@ -27,8 +27,9 @@
 	penetration = ARMOR_PENETRATION_TIER_4
 	accuracy = HIT_ACCURACY_TIER_3
 
-/datum/ammo/bullet/revolver/heavy/on_hit_mob(mob/M, obj/projectile/P)
-	knockback(M, P, 4)
+/datum/ammo/bullet/revolver/heavy/on_hit_mob(mob/entity, obj/projectile/bullet)
+	slowdown(entity, bullet)
+	pushback(entity, bullet)
```
Comment (Zonespace27) [inline code/datums/ammo/bullet/revolver.dm:32]: pushback takes 3 args, readd the 4

**Review by Zonespace27 — APPROVED** [review]

**Pull request merged**